### PR TITLE
shipcat apply without tiller

### LINKED ISSUE #151: Smarter reconciles: Part 2: ownerRefs + kubectl diff
Part two of smarter reconciles. Follows #150.

## Plan:

- Star using ownerRefs in crds in #140
- Stop using `helm upgrade`, pipe the template to `kubectl apply` when the CRD is being configured
- Remove `cluster helm reconcile` command.
- Use `kubectl diff` from `cluster crd reconcile` (already implemented this in `diff.rs`)
- Extend `ReconciliationMode` to have a `CrdOwned` entry to reflect this

Note that `kubectl diff` requires kubernetes server version of 1.13. The enum should allow us to gradually switch over to this mode when our clusters become new enough.

## Pros:

- no tiller, no security elevation craziness, faster upgrades
- automatic deletion of extraneous services - crd reconcile has a tested removal path (that only considers manifests)
- CRD is controlled by `shipcat` so ownerrefs are correct
- no more `helm del --purge` requirements

## Cons:

- no more history of kube secrets (could piggy-back on `helm history` before for this when secrets got changed accidentally)

We can fix this main con by upgrading to vault kv2 #99, which lets you see secrets at older versions.

Once this is done, we can start doing namespaces properly. See part three in #152

### PR DESCRIPTION
**Pull request by clux**
New `ReconciliationMode` feature flagged in `shipcat.conf` per region: `CrdOwned`.

- defines `ownerReferences` in charts so garbage collection works
- calls out to kube to get the uid for the owner refs
- fixes shipcat diff
- implements the new apply function for that mode which now writes a temporary template file
- runs `kubectl apply -f {templatefile} --prune -lapp={myapp}`
- normal watch and status writing from `CrdStatus`

:eyes:  please!

tests in examples/ work with minikube!

Fixes #151 and #11 and #183

### CODE REVIEW

File: examples/charts/base/templates/configmap.yaml [hunk 1]
```diff
@@ -3,16 +3,19 @@ apiVersion: v1
 kind: ConfigMap
 metadata:
   name: {{ .Values.name }}-config
-  namespace: {{ .Release.Namespace }}
+  namespace: {{ .Values.namespace }}
   labels:
     app: {{ .Values.name }}
     type: {{ .Values.type | default "service" }}
 {{- if $.Values.labels }}
 {{ toYaml $.Values.labels | indent 4 }}
```
Comment (jackatbancast) [inline examples/charts/base/templates/configmap.yaml]: This should not be `$.Values.labels`.

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,9 @@
+0.126.0 / 2019-09-XX
```
Comment (gleaver) [inline CHANGELOG.md]: `0.127.0`?

File: examples/charts/base/templates/configmap.yaml [hunk 3]
```diff
@@ -3,16 +3,19 @@ apiVersion: v1
...
 {{- end }}
-    chart: {{ template "chart.chart" . }}
-    release: {{ .Release.Name }}
-    heritage: {{ .Release.Service }}
+  ownerReferences:
+  - apiVersion: babylontech.co.uk/v1
+    kind: ShipcatManifest
+    controller: false
```
Comment (gleaver) [inline examples/charts/base/templates/configmap.yaml]: What does `controller` do?

File: examples/services/blog/manifest.yml [hunk 4]
```diff
@@ -2,10 +2,7 @@ name: blog
 image: clux/blog
 version: 0.1.0
 metadata:
-  contacts:
-  - name: "Eirik"
-    slack: "@clux"
-  team: Doves
+  team: doves
```
Comment (gleaver) [inline examples/services/blog/manifest.yml]: 🕊

File: examples/charts/base/templates/configmap.yaml (hunk 3 above)
Comment (clux) [reply examples/charts/base/templates/configmap.yaml]: it means that kube will delegate to a controller to be in charge of the cleanup rather than kubernetes itself doing it blindly.

File: shipcat_cli/src/apply.rs [hunk 5]
```diff
@@ -509,3 +509,261 @@ pub fn diff_helm(mf: &Manifest, hfile: &str) -> Result<Option<String>> {
...
+    let explicit_version = mfbase.version.clone().or(passed_version);
+
+    // Interact with the kube api to get the shipcatmanifest crd and its .status
+    // This lets us work out:
+    // - if the service has been installed before (negates the need for a diff)
+    // - if we need to apply a new crd (so we have an atomic change)
+    // - if we need to interact with secret-manager TODO: do
+    let s = status::Status::new(&mfbase)?;
+
+    // Next large batch is working out the reason for the upgrade (if any)
```
Comment (gleaver) [inline shipcat_cli/src/apply.rs]: Does it makes sense to break the various stages out into subfunctions?

File: shipcat_cli/src/apply.rs [hunk 6]
```diff
@@ -509,3 +509,261 @@ pub fn diff_helm(mf: &Manifest, hfile: &str) -> Result<Option<String>> {
...
+            Ok(Some(kdiff)) => {
+                ui.diff = Some(kdiff);
+                reason = reason.or(Some(UpgradeReason::TemplateDiff));
+            },
+            Ok(None) => {
+                // If we explicitly received no diff, don't try to upgrade
+                // This is a stronger diff than CRD-only if this succeeds; STOP.
+                info!("{} up to date (full diff check)", svc);
+                s.update_generate_true()?; // every force reconcile makes one generate cond
+                webhooks::apply_event(UpgradeState::Cancelled, &ui, &region, &conf);
```
Comment (gleaver) [inline shipcat_cli/src/apply.rs]: Shouldn't the wehook be before the update_generate in case it fails?

File: shipcat_cli/src/apply.rs [hunk 7]
```diff
@@ -509,3 +509,261 @@ pub fn diff_helm(mf: &Manifest, hfile: &str) -> Result<Option<String>> {
...
+                return Ok(None)
+            }
+            // If diffing failed, only run the upgrade if using --force
+            Err(e) => {
+                warn!("Unable to diff against {}: {}", svc, e);
+                if !force {
+                    // pass on a diff failure
+                    s.update_generate_false("DiffFailure", e.description().to_string())?;
+                    webhooks::apply_event(UpgradeState::Cancelled, &ui, &region, &conf);
```
Comment (gleaver) [inline shipcat_cli/src/apply.rs]: Ditto ordering

File: shipcat_cli/src/apply.rs [hunk 8]
```diff
@@ -509,3 +509,261 @@ pub fn diff_helm(mf: &Manifest, hfile: &str) -> Result<Option<String>> {
...
+                    return Ok(None) // but ultimately ignore this in fast reconciles
+                }
+                reason = reason.or(Some(UpgradeReason::Forced))
+            }
+        }
+    }
+
+    // We cannot be here without a reason now, although you have to convince yourself.
+    let ureason = reason.expect("cannot apply without a reason");
+    s.update_generate_true()?; // if this fails, stop, want .status to be correct
```
Comment (gleaver) [inline shipcat_cli/src/apply.rs]: ordering

File: shipcat_cli/src/apply.rs [hunk 9]
```diff
@@ -509,3 +509,261 @@ pub fn diff_helm(mf: &Manifest, hfile: &str) -> Result<Option<String>> {
...
+}
+
+/// Shell out to kubectl apply
+///
+/// Assumes you have written your template file from `helm template`
+fn upgrade_kubectl(mf: &Manifest, tfile: &str) -> Result<()> {
+    // upgrade it using the same command
+    let applyvec = vec![
+        "apply".into(),
+        format!("-n={}", mf.namespace),
```
Comment (gleaver) [inline shipcat_cli/src/apply.rs]: Possibly paranoia, but do we validate namespace doesn't contain any spaces etc for weird injection problems?

File: shipcat_cli/src/main.rs [hunk 10]
```diff
@@ -599,9 +603,19 @@ fn dispatch_commands(args: &ArgMatches) -> Result<()> {
...
+        let mut mf = if a.is_present("secrets") {
+            shipcat_filebacked::load_manifest(&svc, &conf, &region)?.complete(&region)?
+        } else {
+            shipcat_filebacked::load_manifest(&svc, &conf, &region)?.stub(&region)?
+        };
+        mf.version = mf.version.or(ver);
+        if a.is_present("current") {
+            let s = status::Status::new(&mf)?;
+            let crd = s.get()?;
+            mf.version = crd.spec.version;
```
Comment (gleaver) [inline shipcat_cli/src/main.rs]: shouldn't override version that's running if there's a diffrerent things locally

File: shipcat_cli/src/apply.rs (hunk 5 above)
Comment (clux) [reply shipcat_cli/src/apply.rs]: it makes it harder to reason about the feature gating. this way we can read them top-to-bottom, and while they are big, it's pretty clear to reason about.

in the future, when there's only one reconciliationMode, maybe.

File: shipcat_cli/src/apply.rs (hunk 9 above)
Comment (clux) [reply shipcat_cli/src/apply.rs]: valid paranoia, but this is  a guaranteed output-only property (skip deserialize)

File: shipcat_cli/src/apply.rs [hunk 11]
```diff
@@ -509,3 +508,270 @@ pub fn diff_helm(mf: &Manifest, hfile: &str) -> Result<Option<String>> {
         None
     })
 }
+
+/// First version of apply that does not use tille
```
Comment (jackatbancast) [inline shipcat_cli/src/apply.rs]:
```suggestion
/// First version of apply that does not use tiller
```

**Comment by clux** [comment]
`shipcat template webapp -kc` now validates chart assumptions:

```sh
shipcat::helm: Secret: valid app label webapp
shipcat::helm: Secret: valid namespace apps
shipcat::helm: Secret: valid ownerReference for ShipcatManifest
shipcat::helm: Secret: valid uid in ownerReference for aa01436f-d0b3-11e9-90b9-9cb6d0eb90e9
shipcat::helm: ConfigMap: valid app label webapp
shipcat::helm: ConfigMap: valid namespace apps
shipcat::helm: ConfigMap: valid ownerReference for ShipcatManifest
shipcat::helm: ConfigMap: valid uid in ownerReference for aa01436f-d0b3-11e9-90b9-9cb6d0eb90e9
shipcat::helm: ServiceAccount: valid app label webapp
shipcat::helm: ServiceAccount: valid namespace apps
shipcat::helm: ServiceAccount: valid ownerReference for ShipcatManifest
shipcat::helm: ServiceAccount: valid uid in ownerReference for aa01436f-d0b3-11e9-90b9-9cb6d0eb90e9
shipcat::helm: Service: valid app label webapp
shipcat::helm: Service: valid namespace apps
shipcat::helm: Service: valid ownerReference for ShipcatManifest
shipcat::helm: Service: valid uid in ownerReference for aa01436f-d0b3-11e9-90b9-9cb6d0eb90e9
shipcat::helm: Deployment: valid app label webapp
shipcat::helm: Deployment: valid namespace apps
shipcat::helm: Deployment: valid ownerReference for ShipcatManifest
shipcat::helm: Deployment: valid uid in ownerReference for aa01436f-d0b3-11e9-90b9-9cb6d0eb90e9
```

File: examples/charts/base/templates/configmap.yaml (hunk 3 above)
Comment (janslow) [reply examples/charts/base/templates/configmap.yaml]: Is there a way we could use `kustomize` or similar to inject the owner references (and `app` label) into each object?

File: shipcat_cli/Cargo.toml [hunk 12]
```diff
@@ -31,7 +31,7 @@ clap = "2.33.0"
 error-chain = "0.12.1"
 log = "0.4.5"
 loggerv = "0.7.1"
-kube = "0.15.1"
```
Comment (janslow) [inline shipcat_cli/Cargo.toml]: No `0.16.0`? 😄

File: shipcat_cli/src/apply.rs (hunk 5 above)
Comment (janslow) [reply shipcat_cli/src/apply.rs]: Yeah, it's definitely getting quite hard to follow (as indicated by the need for so many comments)

File: shipcat_cli/Cargo.toml (hunk 12 above)
Comment (clux) [reply shipcat_cli/Cargo.toml]: i released that this this morning and realized it missed a field :<

File: shipcat_cli/src/apply.rs (hunk 5 above)
Comment (clux) [reply shipcat_cli/src/apply.rs]: I can work on breaking it up a bit. :+1:

File: examples/charts/base/templates/configmap.yaml (hunk 3 above)
Comment (clux) [reply examples/charts/base/templates/configmap.yaml]: Maybe. I wasn't planning on touching kustomize in this iteration though.

File: shipcat_cli/src/apply.rs [hunk 13]
```diff
@@ -509,3 +509,250 @@ pub fn diff_helm(mf: &Manifest, hfile: &str) -> Result<Option<String>> {
...
+            },
+        }
+    };
+
+    // Create completed kubernetes yaml (via shipcat values | helm template)
+    let tfile = format!("{}.kube.gen.yml", svc);
+    let tpth = Path::new(".").join(tfile.clone());
+    if let Err(e) = helm::template(&mf,  Some(tpth)) {
+        // Errors here are obscure, and should not happen, but pass them up anyway
+        webhooks::apply_event(UpgradeState::Failed, &ui, &region, &conf);
```
Comment (jackatbancast) [inline shipcat_cli/src/apply.rs]: Not for this PR, but it'd be really useful if we can add a failure reason as well that we can use for debugging.
